# Importing Libraries

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

## Import orders_products_combined.pkl and cleaned products.csv

In [2]:
path = r'C:\Users\Akira\Documents\Instacart Basket Analysis'

In [3]:
df_ords_prods_comb = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

## comparing the two dataframes

In [5]:
# Checking shape - should match 32434489

df_ords_prods_comb.shape

(32434489, 12)

In [6]:
df_prods.shape

(49672, 6)

In [7]:
df_ords_prods_comb.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           0   2539329        1    prior             1          2   
2           0   2539329        1    prior             1          2   
3           0   2539329        1    prior             1          2   
4           0   2539329        1    prior             1          2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8               11.114836         196                  1   
1                  8               11.114836       14084                  2   
2                  8               11.114836       12427                  3   
3                  8               11.114836       26088                  4   
4                  8               11.114836       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [8]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

Looking at the two dataframes, we are able to merge the two data sets using the product_id column as the key 

## Merging the Dataframes

In [9]:
# the indicator= False is because the df_ords_prods_comb is already including a merge flag
df_orders_products_merged = df_ords_prods_comb.merge(df_prods, on = ['product_id'], indicator = False)

In [10]:
df_orders_products_merged

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0                    0   2539329        1    prior             1          2   
1                    1   2398795        1    prior             2          3   
2                    2    473747        1    prior             3          3   
3                    3   2254736        1    prior             4          4   
4                    4    431534        1    prior             5          4   
...                ...       ...      ...      ...           ...        ...   
32404854       3360493   1320836   202557    prior            17          2   
32404855       3360494     31526   202557    prior            18          5   
32404856       3376038    758936   203436    prior             1          2   
32404857       3376039   2745165   203436    prior             2          3   
32404858       3408036   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8               11.114836         196   
1                         7               15.000000         196   
2                        12               21.000000         196   
3                         7               29.000000         196   
4                        15               28.000000         196   
...                     ...                     ...         ...   
32404854                 15                1.000000       43553   
32404855                 11                3.000000       43553   
32404856                  7               11.114836       42338   
32404857                  5               15.000000       42338   
32404858                 14               11.114836       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0_y  \
0                         1          0   both           195   
1                         1          1   both           195   
2                         1          1   both           195   
3                         1          1   both           195   
4                         1          1   both           195   
...                     ...        ...    ...           ...   
32404854                  2          1   both         43557   
32404855                  2          1   both         43557   
32404856                  4          0   both         42342   
32404857                 16          1   both         42342   
32404858                  8          0   both         28820   

                           product_name  aisle_id  department_id  prices  
0                                  Soda        77              7     9.0  
1                                  Soda        77              7     9.0  
2                                  Soda        77              7     9.0  
3                                  Soda        77              7     9.0  
4                                  Soda        77              7     9.0  
...                                 ...       ...            ...     ...  
32404854            Orange Energy Shots        64              7     3.7  
32404855            Orange Energy Shots        64              7     3.7  
32404856          Zucchini Chips, Pesto        50             19     6.9  
32404857          Zucchini Chips, Pesto        50             19     6.9  
32404858  Hot Oatmeal Multigrain Raisin       130             14    10.3  

[32404859 rows x 17 columns]

## Confirming results with merge flag

In [11]:
df_orders_products_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

There is 32404859 rows all in the both category, we can safely use this data for further analysis

## Wrangling the Data - Dropping uncecessary columns

In [12]:
df_orders_products_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             0   2539329        1    prior             1          2   
1             1   2398795        1    prior             2          3   
2             2    473747        1    prior             3          3   
3             3   2254736        1    prior             4          4   
4             4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8               11.114836         196                  1   
1                  7               15.000000         196                  1   
2                 12               21.000000         196                  1   
3                  7               29.000000         196                  1   
4                 15               28.000000         196                  1   

   reordered _merge  Unnamed: 0_y product_name  aisle_id  department_id  \
0          0   both           195         Soda        77              7   
1          1   both           195         Soda        77              7   
2          1   both           195         Soda        77              7   
3          1   both           195         Soda        77              7   
4          1   both           195         Soda        77              7   

   prices  
0     9.0  
1     9.0  
2     9.0  
3     9.0  
4     9.0

The columns Unnamed: 0_x, eval_set,  and Unnamed: 0_y will be removed as they are not necessary for further analysis

In [13]:
df_orders_products_merged2 = df_orders_products_merged.drop(columns = ['Unnamed: 0_x', 'eval_set', 'Unnamed: 0_y'])

In [14]:
df_orders_products_merged2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0               11.114836         196                  1          0   both   
1               15.000000         196                  1          1   both   
2               21.000000         196                  1          1   both   
3               29.000000         196                  1          1   both   
4               28.000000         196                  1          1   both   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

## Checking for missing values

In [15]:
df_orders_products_merged2.isnull().sum()

order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
product_id                0
add_to_cart_order         0
reordered                 0
_merge                    0
product_name              0
aisle_id                  0
department_id             0
prices                    0
dtype: int64

Based on the missing values check, it does not seem like there is any missing values among this dataframe

## Checking for duplicates

In [17]:
#df_dups represents duplicated data
df_dups = df_orders_products_merged2[df_orders_products_merged2.duplicated()]

In [18]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, _merge, product_name, aisle_id, department_id, prices]
Index: []

The duplicates check did not find any duplicate values in the dataframe

In [19]:
df_orders_products_merged2.shape

(32404859, 14)

## Exporting the Data

In [16]:
# Based on the growing size of these files, I will be exporting these files into pkl format as well as csv in order to have both options available

df_orders_products_merged2.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged.csv'))


df_orders_products_merged2.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged.pkl'))